modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridTimestamp.java

```diff
@@ -55,13 +66,35 @@ public final class HybridTimestamp implements Comparable<HybridTimestamp>, Seria
      * @param logical The logical time.
      */
     public HybridTimestamp(long physical, int logical) {
-        assert physical > 0 : physical;
-        // Value -1 is used in "org.apache.ignite.internal.hlc.HybridClock.update" to produce "0" after the increment.
-        // Real usable value cannot be negative.
-        assert logical >= -1 : logical;
+        if (physical < 0 || physical >= (1L << PHYSICAL_TIME_BITS_SIZE)) {
+            throw new IllegalArgumentException("physical time is out of bounds: " + physical);
+        }
 
-        this.physical = physical;
-        this.logical = logical;
+        if (logical < 0 || logical >= (1L << LOGICAL_TIME_BITS_SIZE)) {
+            throw new IllegalArgumentException("logical time is out of bounds: " + logical);
+        }
+
+        time = (physical << LOGICAL_TIME_BITS_SIZE) | logical;
+
+        if (time <= 0) {
+            throw new IllegalArgumentException("time is out of bounds: " + time);
+        }
+    }
+
+    private HybridTimestamp(long time) {
+        if (time <= 0) {
+            throw new IllegalArgumentException("time is out of bounds: " + time);
+        }
+
+        this.time = time;
+    }
+
+    public static @Nullable HybridTimestamp of(long time) {
```

Comment:
Looks like we need to add a bunch of `@Nullable` annotations to methods that use this method's result as their result